### PR TITLE
Extract bounded sanitized display values and escape handling

### PR DESCRIPTION
**Pull request by ronaldmannak**
## Superseded for MVP by #136 and #137

The owner approved structured diagnostics and useful fixed error messages, without arbitrary raw text. #136 (logging) → #137 (typed error/recovery contracts) is the new main-based path. #137 retains the required error categories and OperationID while dropping SanitizedText/MissingComponents/raw associated values. This old sanitizer-dependent error PR is deferred reference, not an MVP merge prerequisite. Its history and unresolved findings remain intact, not declared fixed. Downstream consumers must migrate to #137; do not merge the old redactor ancestry to satisfy their dependencies.

## Scope and dependency

First prerequisite for #53 / issue #7, stacked on #59 (9000cc2). Implements MVP-PLAN.md §3 (Local storage) and supports the §9/§10 typed error boundary.

Provides bounded SanitizedText and its shared scalar/window policy. The escape helper reuses the composed redactor; it does not duplicate the authoritative parser. It retains whole-run concealment for ambiguous terminal splices, recovers edge evidence without emitting control payloads, and preserves ordinary version styling. BEL terminates OSC only; other control strings require ST. MissingComponents belongs to the following presentation prerequisite.

## Latest review repairs

- Open network-path authorities at the inspection cutoff use the supported URI prefix boundaries, including assignment and escaped-slash forms.
- Nested control-payload escapes make recovery ambiguous, so the affected bounded display value is quarantined rather than guessing a terminal interpretation.
- Raw credential delimiters and cursor context are scanned before destructive normalization. Unicode physical separators receive a consistent credential reading.
- An independent normalized reading protects token suffixes that an earlier raw replacement could detach. If that reading discovers credential evidence and disagrees with the raw-derived reading, the whole bounded display value becomes `[redacted:normalized-value]`.
- Truncation accounting uses the pre-redaction normalization length, not shortening introduced by replacement markers.

This is a display-value primitive, not a host operation. Every step remains bounded by the inspection window; the full error serialization and cutoff corpus remains in #53.

## Validation

Current head **f969ee6**: **525 changed lines**, **401 Core tests / 34 suites** pass with warnings-as-errors. The initial four regression functions reproduced 31 failed assertions; Unicode whitespace and competing-readings regressions are retained as well. The original 110 expanded escape-edge cases remain. Complete #53 integration passes 444 tests / 37 suites.

No test methods were removed, no dependencies or host operations added, and no history rewritten. Fresh exact-head Xcode Cloud, standalone/inline Codex review and original-message reaction remain required; old approvals are not current-head approval.

## Repair checkpoint — 2026-09-05 21:14 UTC

Current head `f969ee69c05587b2f16b968fa457b5d6c11ddaa4`: **525 own changed lines; 401 Core tests / 34 suites pass with warnings-as-errors**.

Dropped-scalar artificial boundaries are scanned independently without discarding original cursor evidence. An existing whole-value marker safely keeps its kind; other conflicting readings quarantine the bounded display value. Truncated quoted wrappers are concealed before content is certified.

The complete error graph passes **444 tests / 37 suites**, including unchanged original URI, filename-prefix and combining-mark assertions. The public redactor passes373 tests. Fifteen prerequisite findings received published-fix or evidence-backed policy replies; the public framing, earlier failed-assertion report, two sanitizer boundaries and locale assertion findings were also resolved with evidence.

**Not aggregate merge-ready.** Fresh review still reports three nested-encoding gaps on #119 (Unicode whitespace escapes, mixed raw quote wrappers, parenthesized values), and four public-boundary findings on #59 (validation recovery guidance, decoding scope, split strict credentials, contextual typed output). #59 has more than100 threads; the second page must be inspected. Current-head CI/reviews/reactions remain required after this push; previous green checks and thumbs-up do not clear these findings. No PR was merged.

### CODE REVIEW

File: Packages/GuesthouseCore/Sources/GuesthouseCore/Errors/SanitizedText+Sanitization.swift [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            // below.
+            if normalized.unicodeScalars.count < limit + Self.sanitizeLookahead {
+                normalized = normalized.replacing(#/\S+$/#, with: Redactor.marker("truncated"))
+            }
+            // A URL authority still open at the cut may be userinfo whose terminating `@` fell
+            // outside the window: treat the whole remainder as a credential. The delimiter is
+            // recognized in both the spellings the redactor's own userinfo rule accepts, since a
+            // URL that reached a log through JSON keeps that encoding's escaped slashes, and is
+            // put back exactly as it came in.
+            normalized = normalized.replacing(#/(:(?:\\?\/){2})[^\s\/]*$/#) { match in "\(match.1)\(Redactor.marker("userinfo"))" }
```
[the remark this answers is not in the document]
Comment (ronaldmannak) [reply Packages/GuesthouseCore/Sources/GuesthouseCore/Errors/SanitizedText+Sanitization.swift]: Fixed in 5e7991f (initial repair f3f4ae5). Open-authority cutoff handling now includes the supported network-path leading boundaries, assignment prefixes and escaped slashes, not just ://. Four long synthetic network-path cases reproduced the leak before the repair. #121 passes 370 tests; the full #53 integration passes 413 with warnings-as-errors.

File: Packages/GuesthouseCore/Sources/GuesthouseCore/Redaction/Redactor+Sanitization.swift [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        let scalars = Array(text.unicodeScalars)
+        var runs: [SanitizationRun] = []
+        var run = SanitizationRun()
+        var index = 0
+        while index < scalars.count {
+            if let escape = sanitizationEscape(in: scalars, at: index) {
+                if let payload = escape.payload {
+                    // Remove the entire control string atomically. In particular, a BEL in a
+                    // DCS/APC/PM/SOS payload must not detach its remaining text from its opener.
+                    run.recovered += String(String.UnicodeScalarView(scalars[payload].filter(sanitizationKeepsScalar)))
```
[the remark this answers is not in the document]
Comment (ronaldmannak) [reply Packages/GuesthouseCore/Sources/GuesthouseCore/Redaction/Redactor+Sanitization.swift:18]: Fixed in 5e7991f (initial repair f3f4ae5). Nested ESC/C1 controls make the recovered control-payload reading ambiguous; the bounded display-value sanitizer now quarantines that run and the remainder rather than guessing nested terminal semantics or returning a detached token suffix. Twelve OSC/DCS/ESC/C1/charset cases reproduced the leak and now pass. Existing ordinary terminal-title/version and atomic non-OSC BEL tests remain enabled. #121: 370 passing tests; full #53: 413.

File: Packages/GuesthouseCore/Sources/GuesthouseCore/Redaction/Redactor+Sanitization.swift [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        }
+    }
+
+    /// Matches the error sanitizer's scalar normalization; recovery must inspect the same
+    /// visible characters, while the helper leaves their original rendering in its output.
+    static func sanitizationKeepsScalar(_ scalar: Unicode.Scalar) -> Bool {
+        switch scalar.properties.generalCategory {
+        case .control, .format, .lineSeparator, .paragraphSeparator, .privateUse, .surrogate, .unassigned,
+             .nonspacingMark, .spacingMark, .enclosingMark:
+            false
```
[the remark this answers is not in the document]
Comment (ronaldmannak) [reply Packages/GuesthouseCore/Sources/GuesthouseCore/Redaction/Redactor+Sanitization.swift:113]: Fixed in 5e7991f. Raw credential whitespace is scanned before normalization can join an option/scheme to its value, with explicit Unicode separator handling and cursor-context coverage. The independent normalized reading is also retained: full integration caught a raw-prefix replacement hiding a normalization-assembled suffix, so disagreements involving normalized credential evidence now quarantine the bounded value whole. Tabs, CR/LF/CRLF, VT/FF/NEL, Unicode separators, non-ASCII split-token spaces and mixed raw/normalized credentials have regression coverage. All original no-leak assertions remain; only two marker expectations changed to the documented normalized-value quarantine. #121 has 370 passing tests and #53 has 413.

File: Packages/GuesthouseCore/Sources/GuesthouseCore/Errors/SanitizedText+Sanitization.swift [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        let stripped = Redactor.stripTerminalEscapes(Redactor().redact(fieldValue: credentialReading))
+        var normalized = Redactor().redact(fieldValue: String(String.UnicodeScalarView(stripped.unicodeScalars.filter(Redactor.sanitizationKeepsScalar))))
+        let joinedRedacted = Redactor().redact(fieldValue: joined)
+        // A raw scan may replace a prefix before normalization assembles its secret suffix.
+        // Neither transformed string retains offsets into the other. If the normalized reading
+        // finds credential evidence and disagrees, quarantine this bounded display value whole.
+        if joinedRedacted != joined, joinedRedacted != normalized {
+            normalized = Redactor.marker("normalized-value")
+        }
+        if truncated {
```
[the remark this answers is not in the document]
Comment (ronaldmannak) [reply Packages/GuesthouseCore/Sources/GuesthouseCore/Errors/SanitizedText+Sanitization.swift:68]: Fixed in 3ef8a88. A boundary-delimited still-open Basic value at a truncated inspection window is concealed without requiring the fragment to decode. All four Base64 alignments have regression coverage; the failing 1-modulo-4 case now passes. The inspection budget is unchanged. #121passes382 tests; the complete error stack8bba3fd passes425 with warnings-as-errors.

File: Packages/GuesthouseCore/Sources/GuesthouseCore/Errors/SanitizedText+Sanitization.swift [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            // below.
+            if normalizationShortened {
+                normalized = normalized.replacing(#/\S+$/#, with: Redactor.marker("truncated"))
+            }
+            // A URL authority still open at the cut may be userinfo whose terminating `@` fell
+            // outside the window: treat the whole remainder as a credential. The delimiter is
+            // recognized in both the spellings the redactor's own userinfo rule accepts, since a
+            // URL that reached a log through JSON keeps that encoding's escaped slashes, and is
+            // put back exactly as it came in.
+            normalized = normalized.replacing(#/((?::|^|[\s\u{001F}"'(<\[{]|(?:^|[\s\u{001F}"'(<\[{])(?:--?)?[A-Za-z][A-Za-z0-9_.-]*[ \t]*=[ \t]*)(?:\\?\/){2})[^\s\/?#]*$/#) { match in "\(match.1)\(Redactor.marker("userinfo"))" }
```
[the remark this answers is not in the document]
Comment (ronaldmannak) [reply Packages/GuesthouseCore/Sources/GuesthouseCore/Errors/SanitizedText+Sanitization.swift:84]: Fixed in 3ef8a88. A truncated colon-bearing display value with an unknown remainder is conservatively quarantined because the unseen suffix may establish a DSN; the finite input budget is not extended. A whole open URL authority retains its existing explicit userinfo cutoff repair and safe URI rendering. Bare/assigned/prefixed long DSNs, short ordinary colon diagnostics and the original plain/escaped URL-cutoff assertions all pass. This is an intentional bounded-display privacy tradeoff, not a claim that arbitrary colon prose proves a DSN. Full425-test error suite passes.

File: Packages/GuesthouseCore/Sources/GuesthouseCore/Errors/SanitizedText+Sanitization.swift [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        // fragment behind once the bare control scalars are dropped. A sequence that can only
+        // have borrowed its terminator from the value goes further and takes its whole run
+        // with it, because stripping it would silently repair a credential into something the
+        // patterns below no longer recognize. Combining marks go too: a mark inside a token
+        // would otherwise split it out of the redactor's reach.
+        let spliceSafe = Redactor.redactEscapeSplicedRuns(bounded)
+        let originalStripped = Redactor.stripTerminalEscapes(spliceSafe)
+        let joined = String(String.UnicodeScalarView(originalStripped.unicodeScalars.filter(Redactor.sanitizationKeepsScalar)))
+        let normalizationShortened = joined.unicodeScalars.count < limit + Self.sanitizeLookahead
+        let credentialReading = spliceSafe.replacing(#/[\u{000B}\u{000C}\u{0085}\u{2028}\u{2029}]/#, with: "\n")
```
[the remark this answers is not in the document]
Comment (ronaldmannak) [reply Packages/GuesthouseCore/Sources/GuesthouseCore/Errors/SanitizedText+Sanitization.swift]: Fixed in 3ef8a88. The credential reading maps every whitespace/newline separator to ASCII space, including NBSP/figure space and Basic split by physical separators. If that reading finds no credential, the sanitized original joined reading is retained; otherwise the independent joined reading still quarantines conflicting coverage. Parameterized --password/Bearer/valid-Basic cases cover eleven separator forms. Original ordinary-format, combining-mark, URL-rendering and no-leak assertions remain intact. #121passes382 tests; full #53passes425.

File: Packages/GuesthouseCore/Sources/GuesthouseCore/Redaction/Redactor+Sanitization.swift [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
```
[the remark this answers is not in the document]
Comment (ronaldmannak) [reply Packages/GuesthouseCore/Sources/GuesthouseCore/Redaction/Redactor+Sanitization.swift:113]: Fixed in f969ee69c05587b2f16b968fa457b5d6c11ddaa4, composed through #53 head7ae47af46a94c4b90c4a002874e57a0eb0bd8595. Dropped format/combining scalars get an independent artificial-boundary scan, without destroying original cursor/control evidence. Conflicting readings quarantine the bounded value, while an already whole-value redaction marker safely retains its kind. Added zero-width/bidirectional/combining-mark option regressions. All401 sanitizer tests and444 final tests pass with warnings-as-errors; original combining-mark marker assertions are intact.

File: Packages/GuesthouseCore/Sources/GuesthouseCore/Errors/SanitizedText+Sanitization.swift [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    static let sanitizeLookahead = 512
+
+    static func sanitize(_ value: String, limit: Int = 80) -> String {
+        // Public entry points use this policy, so clamp the bound: an extreme limit
+        // would otherwise overflow the window arithmetic and a negative one would hand
+        // `prefix` an invalid length, trapping instead of returning sanitized text.
+        let limit = min(max(limit, 1), SanitizedText.maximumLimit)
+        // Only the window plus one scalar is ever looked at, so the cost is independent of the
+        // input's size.
+        let (window, truncated) = Self.inspectionWindow(value.unicodeScalars, maximum: limit + Self.sanitizeLookahead)
```
[the remark this answers is not in the document]
Comment (ronaldmannak) [reply Packages/GuesthouseCore/Sources/GuesthouseCore/Errors/SanitizedText+Sanitization.swift:25]: Fixed in f969ee69c05587b2f16b968fa457b5d6c11ddaa4. A truncated quoted diagnostic is quarantined before emitting any bounded decoded prefix; one through three JSON encoding layers now conceal the reported complete embedded password even when the closing wrapper is beyond the inspection budget. Only the existing bounded window is read. All401 sanitizer /444 composed tests pass with warnings-as-errors; original URI and filename-prefix display assertions remain intact.

**Comment by ronaldmannak** [comment]
The owner approved structured diagnostics and useful fixed error messages, without arbitrary raw text. #136 (logging) → #137 (typed error/recovery contracts) is the new main-based path. #137 retains the required error categories and OperationID while dropping SanitizedText/MissingComponents/raw associated values. This old sanitizer-dependent error PR is deferred reference, not an MVP merge prerequisite. Its history and unresolved findings remain intact, not declared fixed. Downstream consumers must migrate to #137; do not merge the old redactor ancestry to satisfy their dependencies.

**Comment by ronaldmannak** [comment]
Closing without merging, completing the previously agreed superseded-error cleanup now that the owner has merged #137. The required typed errors, OperationID and recovery behavior are provided by #137 on main; arbitrary SanitizedText/MissingComponents payloads are outside the approved MVP diagnostics design (ADR 0003). This branch, commits, tests and review history are preserved as reference, and unresolved old findings are not declared fixed. Shared consumers are being migrated deliberately, beginning with #55 replacements #138–#141; do not restore the old sanitizer ancestry to satisfy them.
